### DISCUSSION TITLE
Customize payment method in Checkout Block

### DISCUSSION THREAD

**Discussion by ColonelHathii** [woocommerce/woocommerce#45003] [WooCommerce, Blocks, and Site Editing]
Hi folks 👋 

we have followed the instructions 
https://developer.woocommerce.com/2022/07/07/exposing-payment-options-in-the-checkout-block/ 
and
https://developer.woocommerce.com/2022/05/20/hiding-shipping-and-payment-options-in-the-cart-and-checkout-blocks/
and successfully made one of our payment methods available for the checkout block 🥳  

The payment method offers an optional setting: 
_Deactivate "Ship to different address" if customer choose this payment method._

In the `[woocommerce_checkout]`, if the setting is enabled, we used jQuery to check if our payment method was selected and unchecked the "Ship to a different address" checkbox  and hide it 
![01 Ship to a different address](https://github.com/woocommerce/woocommerce-blocks/assets/143522127/7faf2812-bd73-4495-ab46-de87732e3b5c). 
If a different payment method is selected, the checkbox is shown again.

In the checkout block we want to achieve the same for the checkbox "Use same address for billing".

We could already find out that we can use this code to determine the selected payment method:

```
const { activePaymentMethod } = useSelect( ( select ) => {
	const store = select( PAYMENT_STORE_KEY );
	return {
		activePaymentMethod: store.getActivePaymentMethod(),
	};
} );
```

How can we now achieve that the checkbox is clicked (triggered) when our payment method is selected and hidden? If another payment method is selected, the checkbox should be shown again. Can we use `useShippingAsBilling` from `"@woocommerce/base-context"` for this? Unfortunately we can't import this successfully (`Error: Can't resolve '@woocommerce/base-context/'`).

We would be happy if you could give us some advice. 🙏 
Cheers 🍻

**Comment by opr** [comment]
Hi @ColonelHathii thanks for posting, congrats on getting your payment method integrated with WooCommerce Blocks!
You can uncheck the select box like this:

```ts
const CHECKOUT_STORE_KEY = 'wc/store/checkout';
const { __internalSetUseShippingAsBilling } = useDispatch( CHECKOUT_STORE_KEY );
const { activePaymentMethod } = useSelect( ( select ) => {
	const store = select( PAYMENT_STORE_KEY );
	return {
		activePaymentMethod: store.getActivePaymentMethod(),
	};
} );
if ( activePaymentMethod === YOUR_METHOD_NAME ) {
  __internalSetUseShippingAsBilling( true );
  // Some CSS to hide the box will be necessary here, it's not possible through the data stores/any extensibility API.
}
...
```

I'd recommend saving the value of the checkbox before unsetting it so you can put it back when your payment method is deselected, too. You can use the `getUseShippingAsBilling` selector on the `wc/store/checkout` data store.

**Reply by ColonelHathii** [reply]
Hi @opr 

Thank you very much for your advice. ☘️ 
This works for now.

However, we get a warning in the console:

```
Warning: Cannot update a component (`O`) while rendering a different component (`Unknown`). To locate the bad setState() call inside `Unknown`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
```

Can you do something with this? 
Is it maybe because we just pasted the code at `const Label`?

```jsx
const Label = ( props ) => {

	const CHECKOUT_STORE_KEY = 'wc/store/checkout';
	const { __internalSetUseShippingAsBilling } = useDispatch( CHECKOUT_STORE_KEY );

	const { activePaymentMethod } = useSelect( ( select ) => {
		const store = select( PAYMENT_STORE_KEY );
		return {
			activePaymentMethod: store.getActivePaymentMethod(),
		};
	} );
	if ( activePaymentMethod === 'OUR_PAYMENT_METHOD' ) {
	  __internalSetUseShippingAsBilling( true );
	  // Some CSS to hide the box will be necessary here, it's not possible through the data stores/any extensibility API.
	}

	const { PaymentMethodLabel } = props.components;
	return <PaymentMethodLabel text={ label } />;
};
```

Would it make more sense to insert the code differently? If so: where and how?

Thanks a lot 🙂

**Reply by opr** [reply]
> Is it maybe because we just pasted the code at const Label?

Ah right, maybe it is. My recommendation would be to try moving it to different places outside of your label and see if that works. If you're not in a react component you can use [`dispatch`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#dispatch) and [`select`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#select) instead. `select` has a slightly different API so please check the docs for it!

**Reply by ColonelHathii** [reply]
Thanks to the tips, we were now able to implement it successfully. 🎉
